src/bcc_sensor.c

```diff
@@ -1035,25 +1035,25 @@ BPF_LRU(ip6_cache, struct ip6_key, struct ip_entry);
 static inline bool has_ip_cache(struct ip_key *ip_key, u8 flow)
 {
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 8, 0)
-    struct ip_key *ip_entry = ip_cache.lookup(&ip_key->pid);
```

Comment:
This was correct. Older kernels would lookup key by just only the pid. `struct ip_key` returned from the lookup is then compared with the passed in `struct ip_key`. Perhaps the naming isn't that great. The flow is then checked and updated as needed.

Reply:
As long as `pid` is the first entry in the struct, these are equivalent. Passing `ip_key` is consistent with the `ip_cache` definition.

Reply:
@dheater that might be the case but the rest of code isn' quite right.

The UDP cache for older kernels that don't have LRU hashmaps need the key to be pid, so it can be attempted to be evicted. This also means these older kernels don't have flow level caching unless we created a struct to contain addresses, ports and flow data. So, sadly the UDP level caching for older kernels is one connection per pid, it could be expanded but it would still be really limited.

The UDP event caching on LRU hashmaps is only limited by the number of hashtable entries. So if it the hashtable does get full, the old LRU UDP event may pop back up but it's so "old" relatively speaking it wouldn't matter.

Reply:
The UDP caching for older kernels can be simulated or exercised in newer kernels by enabling only that style of caching, setting it to a non LRU hashtable and making the hashtable size really small.

To test the LRU style of caching, just make the LRU hashtable really small.

Reply:
By older, you mean pre-4.8 where `BPF_LRU` maps to `BPF_HASH`?

https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#2-bpf_hash

Looks to me like it should still work since `BPF_HASH` lets you optionally defined the `key_type`, which we are doing; Defining it to be `struct ip_key`... But I like the suggestion to use a small LRU hashtable and test it.

Reply:
The problem with the older kernels is eviction. Caching is useless if you can't evict it. The simplest way to evict here is when the task exits. We could evict at the thread/pid level instead of the tgid level, that could allow "more" caching.

Reply:
I reverted this change to its previous behavior.